src/libsystemd/sd-device/sd-device.c

```diff
@@ -2472,3 +2472,33 @@ _public_ int sd_device_open(sd_device *device, int flags) {
 
         return TAKE_FD(fd2);
 }
+
+_public_ int sd_device_opendir(sd_device *device, const char *subdir, DIR **ret) {
```

Comment:
hmm, i think we should make this an internal function for now and then maybe expose it as public API later, if we really feel the need to.

I think our primary focus should be exposing "high-level" APIs, i.e. instead of giving people the tools to enumerate directories in sysfs as directories (which would be kinda "low-level"), give them the tools to enumerate sysattrs and child devices (wich would be a lot more "high-level"). The sysattr() enumeration calls we already have in sd_device_get_sysattr_first()/sd_device_get_sysattr_next(). And the child enumeration we have in sd_device_enumerator_add_match_parent().

So I think the udev-builtin_net_id.c should probably be ported to sd_device_enumerator_add_match_parent().

Or maybe we should introduce sd_device_get_child_first()/sd_device_get_child_next()? Would return an sd_device, like sd_device_new() would. Major difference to the sd_device_enumerator_add_match_parent() thing would be that it would not be recursive, but just return the first level of sub-devices. I think that would be kinda nice: if you want to enumerate information closely related to a device, you'd sd_device_get_*_first()/_next() and if you want something broader than that use sd_device_enumerator().

Does that make sense?

Reply:
Thanks. Yeah, that makes sense for me.

I will try to rewrite the logic with sd-device-enumerator or so if it possible. But is it OK to leave it to another PR?
In this PR, I'd like to keep the current logic as is, that is, only replace `sd_device_get_syspath() + path_join() + opendir()` with `device_opendir()`.